### PR TITLE
Rectify `native_batch_norm` schema by splitting it into two legit schemas

### PR DESCRIPTION
**Pull request by janeyx99**
Using the same repro from the issue (but with BatchNorm2D)

Rectifies native_batch_norm schema by splitting the schema into 2:
1. one will have NON-optional alias-able running_mean and running_var inputs
2. the other will just not have those parameters at all (no_stats variation)

**Calling for name suggestions!**

## test plan
I've added tests in test_functionalization.py as well as an entry in common_method_invocations.py for `native_batch_norm_legit`
CI should pass.

## next steps
Because of bc/fc reasons, we reroute native_batch_norm to call our new schemas ONLY through the python dispatcher, but in 2 weeks or so, we should make `native_batch_norm_legit` the official batch_norm.

cc @VitalyFedyunin @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @mlazos @soumith @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @Guobing-Chen @chunyuan-w @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @desertfire

### CODE REVIEW

File: aten/src/ATen/native/native_functions.yaml [hunk 1]
```diff
@@ -3811,6 +3811,34 @@
     MPS: batch_norm_mps_out
     CPU: batch_norm_cpu_out
 
+- func: native_batch_norm_legit(Tensor input, Tensor? weight, Tensor? bias, Tensor(a!) running_mean, Tensor(b!) running_var, bool training, float momentum, float eps) -> (Tensor, Tensor, Tensor)
```
Comment (albanD) [inline aten/src/ATen/native/native_functions.yaml:3814]: Can you add a comment with a TODO: Change this to do ... in 2 weeks.

File: test/test_meta.py [hunk 2]
```diff
@@ -519,6 +519,8 @@ def run_meta_crossref(
                     # This should be the same set of kernels
                     # that fake tensor runs in dynamic shapes mode.
                     with enable_python_dispatcher():
+                        print("meta args", meta_args)
```
Comment (albanD) [inline test/test_meta.py:522]: debug prints

File: torch/_decomp/decompositions.py [hunk 3]
```diff
@@ -1373,6 +1373,81 @@ def native_batch_norm(
...
+    input: Tensor,
+    weight: Optional[Tensor],
+    bias: Optional[Tensor],
+    running_mean: Optional[Tensor],
+    running_var: Optional[Tensor],
+    training: bool,
+    momentum: float,
+    eps: float,
+) -> Tuple[Tensor, Tensor, Tensor]:
+    if running_mean is not None and running_var is not None:
```
Comment (albanD) [inline torch/_decomp/decompositions.py:1397]: You should error out if only one of them is provided instead of ignoring it I think.

File: aten/src/ATen/native/native_functions.yaml (hunk 1 above)
Comment (albanD) [inline aten/src/ATen/native/native_functions.yaml:3814]: Also, we should prepend an `_` to make this private.

File: test/test_functionalization.py [hunk 4]
```diff
@@ -1212,5 +1213,117 @@ def forward(self, a_1):
...
+    mean_1 = torch.ops.aten.mean.dim(view_4, [0]);  view_4 = None
+    copy_1 = torch.ops.aten.copy_.default(alias_1, mean_1);  alias_1 = mean_1 = None
+    view_5 = torch.ops.aten.view.default(getitem, [20, 100, 35, 45]);  getitem = None
+    return view_5
+    """)
+
+    def test_batch_norm(self):
+        def f(x):
+            with enable_python_dispatcher():
+                return torch.batch_norm(x, None, None, torch.zeros(100), torch.ones(100), False, 0.1, 1e-5, False)
```
Comment (albanD) [inline test/test_functionalization.py:1289]: Would be good to have another test with f2 that passes x as one of the inputs being modified inplace.

File: torch/_decomp/decompositions.py [hunk 5]
```diff
@@ -1373,6 +1373,81 @@ def native_batch_norm(
...
+    bias: Optional[Tensor],
+    running_mean: Tensor,
+    running_var: Tensor,
+    training: bool,
+    momentum: float,
+    eps: float,
+) -> Tuple[Tensor, Tensor, Tensor, Tensor, Tensor]:
+    new_running_mean = running_mean.clone()
+    new_running_var = running_var.clone()
+    output, save_mean, save_rstd = native_batch_norm(
```
Comment (albanD) [inline torch/_decomp/decompositions.py:1445]: We should refactor the native_batch_norm decomp to be able to choose if mean/rstd are modified inplace or not such that this decomp NEVER introduces new inplace ops.

File: test/test_meta.py (hunk 2 above)
Comment (janeyx99) [reply test/test_meta.py:522]: oop

File: torch/_decomp/decompositions.py [hunk 6]
```diff
@@ -1322,16 +1321,22 @@ def native_batch_norm(
...
         output, mean, rstd = normalize(input, reduction_dims, eps)
 
         save_mean = _squeeze_multiple(mean, reduction_dims)
         save_rstd = _squeeze_multiple(rstd, reduction_dims)
         if running_mean is not None:
-            running_mean.copy_(momentum * save_mean + (1 - momentum) * running_mean)
+            if functional:
+                new_running_mean = momentum * save_mean + (1 - momentum) * running_mean
+            else:
+                running_mean.copy_(momentum * save_mean + (1 - momentum) * running_mean)
```
Comment (albanD) [inline torch/_decomp/decompositions.py:1339]:
```suggestion
            new_running_mean = momentum * save_mean + (1 - momentum) * running_mean
            if not functional:
                running_mean.copy_(new_running_mean)
```

File: torch/_decomp/decompositions.py [hunk 7]
```diff
@@ -1340,7 +1345,12 @@ def native_batch_norm(
             unbiased_var = torch.var(input, reduction_dims, unbiased=False) * (
                 n / (n - 1)
             )
-            running_var.copy_(momentum * unbiased_var + (1 - momentum) * running_var)
+            if functional:
+                new_running_var = momentum * unbiased_var + (1 - momentum) * running_var
```
Comment (albanD) [inline torch/_decomp/decompositions.py:1349]: remove duplication as above.

File: torch/_decomp/decompositions.py [hunk 8]
```diff
@@ -1370,9 +1380,123 @@ def native_batch_norm(
...
+    running_mean: Optional[Tensor],
+    running_var: Optional[Tensor],
+    training: bool,
+    momentum: float,
+    eps: float,
+) -> Tuple[Tensor, Tensor, Tensor]:
+    output, save_mean, save_rstd, _, _ = native_batch_norm_helper(
+        input, weight, bias, running_mean, running_var, training, momentum, eps, False
+    )
+    return output.to(dtype=input.dtype), save_mean, save_rstd
```
Comment (albanD) [inline torch/_decomp/decompositions.py:1406]: double to

**Review by albanD — APPROVED** [review]
Ok!

File: test/test_functionalization.py [hunk 9]
```diff
@@ -1228,6 +1228,205 @@ def forward(self, a_1):
...
+
+    def test_instance_norm_running_mean_is_x(self):
+        def f(x):
+            with enable_python_dispatcher():
+                return torch.instance_norm(torch.randn(20, 100, 35, 45), None, None, running_mean=x, running_var=torch.ones(100),
+                                           use_input_stats=True, momentum=0.1, eps=1e-5, cudnn_enabled=False)
+        # TODO: uncomment following line after functionalization can handle input mutations
+        # self.assert_functionalization(f, torch.zeros(100))
+        logs = self.get_logs(f, torch.zeros(100))
+        # On Windows, for instance_norm, the alias_copy's are reordered to come right before they need to be used
```
Comment (janeyx99) [inline test/test_functionalization.py:1315]: @bdhirsh Want to confirm that this is an okay explanation to pass the failing Windows test

Comment (janeyx99) [reply test/test_functionalization.py:1315]: image
For reference, the log checking failed on Windows as alias_copy calls were ordered differently.

Comment (bdhirsh) [reply test/test_functionalization.py:1315]: yep this looks fine!

File: torch/_decomp/decompositions.py [hunk 10]
```diff
@@ -1370,7 +1378,127 @@ def native_batch_norm(
     if input.device.type == "cpu":
         save_mean = save_mean.to(dtype=input.dtype)
         save_rstd = save_rstd.to(dtype=input.dtype)
-    return output.to(dtype=input.dtype), save_mean, save_rstd
+    return (
```
Comment (bdhirsh) [inline torch/_decomp/decompositions.py:1381]: nit: the 2 extra returns are marked optional, but you always return them.

Comment (janeyx99) [reply torch/_decomp/decompositions.py:1381]: Doesn't the optional mean that I return None instead of a Tensor sometimes? (Which is true)

Comment (bdhirsh) [reply torch/_decomp/decompositions.py:1381]: It looks like in the case when `training=False,` we always set new_running_mean/new_running_var to be the current running_mean/running_var, and it's possible for `train` to be False, but for the user to still pass in a non-None running_mean/running_var.

This is fine to clean up later though: maybe ensuring that we always return None, and then in the non-functional decomps that call this helper, asserting that the last two outputs are None.
